File: tests/fixtures/user/linux/src/mlock.c
/* Memory locking (mm/mlock.c, mm/gup.c, mm/mmap.c, mm/mremap.c,
 * mm/madvise.c): mlock, mlock2, and munlock (the right, RLIMIT_MEMLOCK
 * less what a range already holds locked, holes, PROT_NONE, whole pages,
 * the kernel's length arithmetic), populating or not (MLOCK_ONFAULT),
 * madvise's refusals, mlockall with MCL_FUTURE for mmap, brk, and mremap,
 * MAP_LOCKED, VmLck in /proc/self/status, and a child process, which
 * inherits no lock. Root drops to nobody first; only the soft limit is
 * changed. */
#define _GNU_SOURCE
#include <fcntl.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/mman.h>
#include <sys/resource.h>
#include <sys/syscall.h>
#include <sys/wait.h>
#include <unistd.h>
#include "check.h"

#define P 4096UL
#ifndef MLOCK_ONFAULT
#define MLOCK_ONFAULT 1
#endif
#ifndef MCL_ONFAULT
#define MCL_ONFAULT 4
#endif
#ifndef MADV_COLD
#define MADV_COLD 20
#endif
#ifndef MADV_DONTNEED_LOCKED
#define MADV_DONTNEED_LOCKED 24
#endif

/* VmLck in kB, -1 if missing. */
static long vmlck(void) {
    static char b[8192];
    int fd = open("/proc/self/status", O_RDONLY);
    long n = fd >= 0 ? read(fd, b, sizeof b - 1) : -1;
    if (fd >= 0) close(fd);
    if (n <= 0) return -1;
    b[n] = 0;
    char *l = strstr(b, "\nVmLck:");
    return l ? atol(l + 7) : -1;
}

static int resident(void *a) {
    unsigned char v = 0;
    return mincore(a, P, &v) == 0 && (v & 1);
}

static int memlock(unsigned long pages) {
    struct rlimit r;
    getrlimit(RLIMIT_MEMLOCK, &r);
    r.rlim_cur = pages * P;
    return setrlimit(RLIMIT_MEMLOCK, &r);
}

static char *map(size_t len, int prot) {
    return mmap(NULL, len, prot, MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
}

static void locking(void) {
    char *m = map(32 * P, PROT_READ | PROT_WRITE);
    CHECK_ERR("mlock2-flags", syscall(SYS_mlock2, m, P, 2), EINVAL);
    CHECK("limit-0", memlock(0) == 0);
    CHECK_ERR("mlock-no-right", mlock(m, P), EPERM);
    CHECK_ERR("mlockall-no-right", mlockall(MCL_CURRENT), EPERM);
    /* musl's mmap turns this EPERM into ENOMEM; the call itself: */
    CHECK_ERR("map-locked-no-right",
              syscall(SYS_mmap, NULL, P, PROT_READ, MAP_PRIVATE | MAP_ANONYMOUS | MAP_LOCKED, -1, 0),
              EPERM);
    CHECK("munlock-no-right-needed", munlock(m, P) == 0);
    CHECK("limit-16", memlock(16) == 0);
    CHECK("vmlck-none", vmlck() == 0);
    CHECK("mlock-8", mlock(m, 8 * P) == 0 && vmlck() == 32);
    CHECK_ERR("mlock-over-limit", mlock(m + 8 * P, 16 * P), ENOMEM);
    CHECK("mlock-again", mlock(m, 8 * P) == 0 && vmlck() == 32);
    CHECK("mlock-overlap-counted-once", mlock(m + 4 * P, 12 * P) == 0 && vmlck() == 64);
    CHECK("munlock-whole-pages", munlock(m + 2 * P + 1, P) == 0 && vmlck() == 56);
    CHECK("mlock-size-max", mlock(m + 20 * P, (size_t)-1) == 0 && vmlck() == 56);
    CHECK("munlock-all-of-it", munlock(m, 32 * P) == 0 && vmlck() == 0);
    /* Holes, PROT_NONE. */
    munmap(m + 25 * P, P);
    CHECK_ERR("mlock-hole-first", mlock(m + 25 * P, 2 * P), ENOMEM);
    CHECK("mlock-hole-first-nothing", vmlck() == 0);
    CHECK_ERR("mlock-hole-later", mlock(m + 24 * P, 3 * P), ENOMEM);
    CHECK("mlock-hole-later-before", vmlck() == 4);
    mprotect(m + 28 * P, P, PROT_NONE);
    CHECK_ERR("mlock-prot-none", mlock(m + 28 * P, P), ENOMEM);
    CHECK("mlock-prot-none-locked", vmlck() == 8);
    CHECK("munlock-holes", munlock(m + 24 * P, P) == 0 && munlock(m + 28 * P, P) == 0 &&
                               vmlck() == 0);
    /* Bringing pages in. */
    char *f = map(4 * P, PROT_READ | PROT_WRITE);
    CHECK("onfault", syscall(SYS_mlock2, f, 2 * P, MLOCK_ONFAULT) == 0 && !resident(f) &&
                         vmlck() == 8);
    CHECK("populated", mlock(f + 2 * P, 2 * P) == 0 && resident(f + 3 * P));
    char *w = map(P, PROT_WRITE), *x = map(P, PROT_EXEC);
    CHECK("write-only-populated", mlock(w, P) == 0 && resident(w));
    CHECK("exec-only-populated", mlock(x, P) == 0 && resident(x));
    /* Locked pages are not discarded. */
    memset(m, 7, 2 * P);
    CHECK("lock-for-advice", mlock(m, P) == 0);
    CHECK_ERR("dontneed-locked", madvise(m, 2 * P, MADV_DONTNEED), EINVAL);
    CHECK_ERR("free-locked", madvise(m, 2 * P, MADV_FREE), EINVAL);
    CHECK_ERR("cold-locked", madvise(m, 2 * P, MADV_COLD), EINVAL);
    CHECK("kept", m[0] == 7);
    CHECK("dontneed-locked-ok", madvise(m, P, MADV_DONTNEED_LOCKED) == 0 && m[0] == 0);
    CHECK("munlock-then-dontneed", munlock(m, P) == 0 && madvise(m + P, P, MADV_DONTNEED) == 0 &&
                                       m[P] == 0);
    munlockall();
    CHECK("munlockall", vmlck() == 0);
    munmap(m, 32 * P);
    munmap(f, 4 * P);
}

static void future(void) {
    CHECK_ERR("mlockall-none", mlockall(0), EINVAL);
    CHECK_ERR("mlockall-unknown", mlockall(8), EINVAL);
    CHECK_ERR("mlockall-onfault-alone", mlockall(MCL_ONFAULT), EINVAL);
    CHECK_ERR("mlockall-current-too-big", mlockall(MCL_CURRENT), ENOMEM);
    CHECK("mlockall-future", mlockall(MCL_FUTURE) == 0 && vmlck() == 0);
    char *m = map(4 * P, PROT_READ | PROT_WRITE);
    CHECK("future-mapping-locked", m != MAP_FAILED && vmlck() == 16 && resident(m + 3 * P));
    CHECK("future-over-limit", map(16 * P, PROT_READ) == MAP_FAILED && errno == EAGAIN);
    char *brk = (char *)syscall(SYS_brk, 0);
    char *top = (char *)(((unsigned long)brk + P - 1) & ~(P - 1));
    CHECK("brk-locked", (char *)syscall(SYS_brk, top + 2 * P) == top + 2 * P && vmlck() == 24);
    CHECK("brk-over-limit", (char *)syscall(SYS_brk, top + 64 * P) == top + 2 * P);
    char *g = mremap(m, 4 * P, 6 * P, MREMAP_MAYMOVE);
    CHECK("mremap-grow-locked", g != MAP_FAILED && vmlck() == 32 && resident(g + 5 * P));
    CHECK("mremap-over-limit", mremap(g, 6 * P, 16 * P, MREMAP_MAYMOVE) == MAP_FAILED &&
                                   errno == EAGAIN);
    /* MREMAP_DONTUNMAP unlocks the old range but leaves it counted:
     * move_vma counts the new VMA and never unmaps the old one. */
    char *d = mremap(g, 6 * P, 6 * P, MREMAP_MAYMOVE | MREMAP_DONTUNMAP);
    CHECK("dontunmap-counted-twice", d != MAP_FAILED && vmlck() == 56 && !resident(g));
    /* A child starts unlocked, MCL_FUTURE and all. */
    fflush(stdout);
    pid_t c = fork();
    if (c == 0) {
        char *n = map(P, PROT_READ | PROT_WRITE);
        _exit(vmlck() == 0 && n != MAP_FAILED && vmlck() == 0 ? 0 : 1);
    }
    int st = 0;
    waitpid(c, &st, 0);
    CHECK("child-unlocked", WIFEXITED(st) && WEXITSTATUS(st) == 0);
    CHECK("munlockall-keeps-the-count", munlockall() == 0 && vmlck() == 24);
    char *n = map(P, PROT_READ | PROT_WRITE);
    CHECK("no-future-lock", n != MAP_FAILED && vmlck() == 24);
    CHECK("map-locked", mmap(NULL, 2 * P, PROT_READ, MAP_PRIVATE | MAP_ANONYMOUS | MAP_LOCKED, -1,
                             0) != MAP_FAILED &&
                            vmlck() == 32);
}

int main(void) {
    fflush(stdout);
    pid_t c = fork();
    if (c == 0) {
        if (getuid() == 0 && (setgid(65534) != 0 || setuid(65534) != 0)) {
            printf("FAIL drop-privileges\n");
            exit(1);
        }
        locking();
        future();
        fflush(stdout);
        exit(failures ? 1 : 0);
    }
    int status = 0;
    waitpid(c, &status, 0);
    failures = !WIFEXITED(status) || WEXITSTATUS(status) != 0;
    FINISH();
}

File: src/user/linux/tests/mlock.rs
//! Memory locking against `mm/mlock.c`, `mm/gup.c`, `mm/mmap.c`,
//! `mm/mremap.c`, and `mm/madvise.c` (Linux 6.19), through the system calls
//! on every ABI: `mlock`, `mlock2`, and `munlock` (the right, the limit
//! less what is already locked, holes, `PROT_NONE`, the kernel's length
//! arithmetic), populating or not (`MLOCK_ONFAULT`), `mlockall` and
//! `MCL_FUTURE` for `mmap`, `brk`, and `mremap`, `MAP_LOCKED`, `madvise`'s
//! refusals, `VmLck`, and a new process.

use super::harness::{Harness, P, each_abi};
use crate::user::linux::abi::errno_table::*;
use crate::user::linux::abi::{LinuxAbi, Sysno, vma_flags};

const NOBODY: u32 = 65534;
const RW: u64 = 3;
const MAP_PRIVATE_ANON: u64 = 0x22;
const MAP_LOCKED: u64 = 0x2000;
const MLOCK_ONFAULT: u64 = 1;
const MCL_CURRENT: u64 = 1;
const MCL_FUTURE: u64 = 2;
const MCL_ONFAULT: u64 = 4;
const MADV_DONTNEED: u64 = 4;
const MADV_FREE: u64 = 8;
const MADV_COLD: u64 = 20;
const MADV_DONTNEED_LOCKED: u64 = 24;
const MREMAP_MAYMOVE: u64 = 1;
const MREMAP_DONTUNMAP: u64 = 4;

fn creds(h: &mut Harness, id: u32) {
    h.proc.state.creds = (id, id, id, id);
    h.proc.state.groups = Vec::new();
}

/// `RLIMIT_MEMLOCK` of `pages` pages.
fn memlock(h: &mut Harness, pages: u64) {
    h.proc.state.rlimits[8] = (pages * P, pages * P);
}

/// `mm->locked_vm`, in pages.
fn locked(h: &Harness) -> u64 {
    crate::user::linux::syscall::mlock::locked_pages(&h.proc.state)
}

fn is_locked(h: &Harness, addr: u64) -> bool {
    let v = h.proc.state.space.vma_at(addr).unwrap();
    v.flags & vma_flags::LOCKED != 0
}

#[test]
fn mlock_checks_the_right_the_limit_and_the_range() {
    each_abi(|abi| {
        let mut h = Harness::new(abi);
        creds(&mut h, NOBODY);
        let m = h.anon(32 * P, RW, false);
        assert_eq!(h.err(Sysno::Mlock2, &[m, P, 2]), EINVAL);
        // can_do_mlock: no limit and no CAP_IPC_LOCK.
        memlock(&mut h, 0);
        assert_eq!(h.err(Sysno::Mlock, &[m, P]), EPERM);
        assert_eq!(h.err(Sysno::Mlockall, &[MCL_CURRENT]), EPERM);
        let fixed = [0, P, RW, MAP_PRIVATE_ANON | MAP_LOCKED, u64::MAX, 0];
        assert_eq!(h.err(Sysno::Mmap, &fixed), EPERM);
        assert_eq!(h.call(Sysno::Munlock, &[m, P]), 0, "no right needed");
        // The limit counts what the range does not already hold locked.
        memlock(&mut h, 16);
        assert_eq!(h.call(Sysno::Mlock, &[m, 8 * P]), 0);
        assert_eq!(locked(&h), 8);
        assert_eq!(h.err(Sysno::Mlock, &[m + 8 * P, 16 * P]), ENOMEM);
        assert_eq!(h.call(Sysno::Mlock, &[m, 8 * P]), 0, "already locked");
        assert_eq!(h.call(Sysno::Mlock, &[m + 4 * P, 12 * P]), 0);
        assert_eq!(locked(&h), 16);
        // Whole pages around the range.
        assert_eq!(h.call(Sysno::Munlock, &[m + 2 * P + 1, P]), 0);
        assert_eq!(locked(&h), 14);
        assert!(!is_locked(&h, m + 3 * P) && is_locked(&h, m + 4 * P));
        // The kernel's arithmetic: SIZE_MAX pages round to nothing.
        assert_eq!(h.call(Sysno::Mlock, &[m + 20 * P, u64::MAX]), 0);
        assert!(!is_locked(&h, m + 20 * P));
        // Holes: none at the start, then ENOMEM after the VMAs before it.
        creds(&mut h, 0);
        h.ok(Sysno::Munmap, &[m + 25 * P, P]);
        assert_eq!(h.err(Sysno::Mlock, &[m + 25 * P, 2 * P]), ENOMEM);
        assert!(!is_locked(&h, m + 26 * P));
        assert_eq!(h.err(Sysno::Mlock, &[m + 24 * P, 3 * P]), ENOMEM);
        assert!(is_locked(&h, m + 24 * P) && !is_locked(&h, m + 26 * P));
        // PROT_NONE is locked but cannot be brought in: ENOMEM.
        h.ok(Sysno::Mprotect, &[m + 28 * P, P, 0]);
        let before = locked(&h);
        assert_eq!(h.err(Sysno::Mlock, &[m + 28 * P, P]), ENOMEM);
        assert_eq!(locked(&h), before + 1);
    });
}

#[test]
fn locking_brings_pages_in_unless_on_fault() {
    each_abi(|abi| {
        let mut h = Harness::new(abi);
        let m = h.anon(4 * P, RW, false);
        let space = &h.proc.state.space;
        assert!(!space.is_resident(m));
        h.ok(Sysno::Mlock2, &[m, 2 * P, MLOCK_ONFAULT]);
        assert!(!h.proc.state.space.is_resident(m), "on fault");
        assert_eq!(locked(&h), 2);
        h.ok(Sysno::Mlock, &[m + 2 * P, 2 * P]);
        assert!(h.proc.state.space.is_resident(m + 3 * P));
        // A write-only or execute-only mapping is still read in.
        let w = h.anon(P, 2, false);
        h.ok(Sysno::Mlock, &[w, P]);
        assert!(h.proc.state.space.is_resident(w));
        let x = h.anon(P, 4, false);
        h.ok(Sysno::Mlock, &[x, P]);
        assert!(h.proc.state.space.is_resident(x));
    });
}

#[test]
fn locked_pages_are_not_discarded() {
    // madvise_dontneed_free_valid_vma, can_madv_lru_vma.
    each_abi(|abi| {
        let mut h = Harness::new(abi);
        let m = h.anon(2 * P, RW, false);
        h.fill(m, 2 * P, 7);
        h.ok(Sysno::Mlock, &[m, P]);
        for advice in [MADV_DONTNEED, MADV_FREE, MADV_COLD] {
            assert_eq!(h.err(Sysno::Madvise, &[m, 2 * P, advice]), EINVAL);
        }
        assert_eq!(h.byte(m), 7);
        h.ok(Sysno::Madvise, &[m, P, MADV_DONTNEED_LOCKED]);
        assert_eq!(h.byte(m), 0);
        h.ok(Sysno::Munlock, &[m, P]);
        h.ok(Sysno::Madvise, &[m + P, P, MADV_DONTNEED]);
        assert_eq!(h.byte(m + P), 0);
    });
}

#[test]
fn mlockall_locks_what_is_and_what_will_be() {
    each_abi(|abi| {
        let mut h = Harness::new(abi);
        creds(&mut h, NOBODY);
        memlock(&mut h, 16);
        for bad in [0, 8, MCL_ONFAULT] {
            assert_eq!(h.err(Sysno::Mlockall, &[bad]), EINVAL, "{bad}");
        }
        assert_eq!(h.err(Sysno::Mlockall, &[MCL_CURRENT]), ENOMEM, "too big");
        h.ok(Sysno::Mlockall, &[MCL_FUTURE]);
        assert_eq!(locked(&h), 0);
        let m = h.anon(4 * P, RW, false);
        assert!(is_locked(&h, m));
        assert!(h.proc.state.space.is_resident(m + 3 * P));
        let big = [0, 16 * P, RW, MAP_PRIVATE_ANON, u64::MAX, 0];
        assert_eq!(h.err(Sysno::Mmap, &big), EAGAIN);
        // The break grows locked, within the limit.
        let brk = h.call(Sysno::Brk, &[0]) as u64;
        let top = (brk + P - 1) & !(P - 1);
        assert_eq!(h.call(Sysno::Brk, &[top + 2 * P]) as u64, top + 2 * P);
        assert!(is_locked(&h, top));
        assert_eq!(h.call(Sysno::Brk, &[top + 64 * P]) as u64, top + 2 * P);
        // mremap: growth within the limit, then beyond it.
        let grown = h.ok(Sysno::Mremap, &[m, 4 * P, 6 * P, MREMAP_MAYMOVE, 0]);
        assert_eq!(locked(&h), 8);
        assert!(h.proc.state.space.is_resident(grown + 5 * P));
        assert_eq!(
            h.err(Sysno::Mremap, &[grown, 6 * P, 16 * P, MREMAP_MAYMOVE, 0]),
            EAGAIN
        );
        // MREMAP_DONTUNMAP leaves the old range unlocked but still counted
        // (move_vma counts the new VMA and never unmaps the old one).
        let moved = h.ok(
            Sysno::Mremap,
            &[grown, 6 * P, 6 * P, MREMAP_MAYMOVE | MREMAP_DONTUNMAP, 0],
        );
        assert!(is_locked(&h, moved) && !is_locked(&h, grown));
        assert_eq!(locked(&h), 14);
        // munlockall ends both, the count keeping what the move left.
        h.ok(Sysno::Munlockall, &[]);
        assert_eq!(locked(&h), 6);
        let n = h.anon(P, RW, false);
        assert!(!is_locked(&h, n));
        // MCL_CURRENT locks every VMA but a special mapping.
        creds(&mut h, 0);
        h.ok(Sysno::Mlockall, &[MCL_CURRENT | MCL_ONFAULT]);
        for v in h.proc.state.space.vma_snapshot() {
            let special = v.flags & vma_flags::SPECIAL != 0;
            assert_eq!(v.flags & vma_flags::LOCKED != 0, !special, "{v:?}");
        }
        assert_eq!(
            abi != LinuxAbi::X86_64,
            h.proc
                .state
                .space
                .vma_snapshot()
                .iter()
                .any(|v| v.name.as_deref() == Some("[vdso]")),
        );
    });
}

#[test]
fn a_special_mapping_is_not_resized_or_kept() {
    // check_prep_vma: VM_DONTEXPAND.
    for abi in [LinuxAbi::Aarch64, LinuxAbi::Riscv64] {
        let mut h = Harness::new(abi);
        let vdso = h
            .proc
            .state
            .space
            .vma_snapshot()
            .into_iter()
            .find(|v| v.name.as_deref() == Some("[vdso]"))
            .unwrap()
            .start;
        let grow = [vdso, P, 2 * P, MREMAP_MAYMOVE, 0];
        assert_eq!(h.err(Sysno::Mremap, &grow), EFAULT);
        let keep = [vdso, P, P, MREMAP_MAYMOVE | MREMAP_DONTUNMAP, 0];
        assert_eq!(h.err(Sysno::Mremap, &keep), EINVAL);
        h.ok(Sysno::Mlock, &[vdso, P]);
        assert!(!is_locked(&h, vdso));
    }
}

#[test]
fn status_shows_vmlck_and_a_new_process_has_none() {
    each_abi(|abi| {
        let mut h = Harness::new(abi);
        let m = h.anon(3 * P, RW, false);
        h.ok(Sysno::Mlock, &[m, 3 * P]);
        let status = crate::user::linux::procfs::status(&h.proc.state, &h.proc.threads[0], 1);
        let text = String::from_utf8(status).unwrap();
        assert!(text.contains("\nVmLck:\t      12 kB\n"), "{text}");
        h.ok(Sysno::Mlockall, &[MCL_FUTURE]);
        crate::user::linux::syscall::mlock::forked(&mut h.proc.state);
        assert_eq!(locked(&h), 0);
        let n = h.anon(P, RW, false);
        assert!(!is_locked(&h, n), "MCL_FUTURE is not inherited");
    });
}
